lib/mox/server.ex

```diff
@@ -0,0 +1,86 @@
+defmodule Mox.Server do
+  @moduledoc false
+
+  use GenServer
+
+  # Public API
+
+  def start_link(_options) do
+    GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
+  end
+
+  def add_expectation(owner_pid, key, {n, calls, stub}) do
+    GenServer.call(__MODULE__, {:add_expectation, owner_pid, key, {n, calls, stub}})
+  end
+
+  def get_expectation(owner_pid, key) do
+    GenServer.call(__MODULE__, {:get_expectation, owner_pid, key})
+  end
+
+  def get_fun_to_call(owner_pid, key) do
+    GenServer.call(__MODULE__, {:get_fun_to_call, owner_pid, key})
+  end
+
+  def keys(owner_pid) do
+    GenServer.call(__MODULE__, {:keys, owner_pid})
+  end
+
+  # Callbacks
+
+  def handle_call({:add_expectation, owner_pid, key, {n, calls, stub}}, _from, state) do
+    new_state = state
+                |> maybe_add_pid(owner_pid)
+                |> update_in([owner_pid, key], &(do_add_expectation(&1, {n, calls, stub})))
+
+    {:reply, :ok, new_state}
+  end
+
+  def handle_call({:get_expectation, owner_pid, key}, _from, state) do
+    {:reply, state[owner_pid][key], state}
+  end
+
+  def handle_call({:get_fun_to_call, owner_pid, key}, _from, state) do
+    case Map.get(state, owner_pid, %{})[key] do
+      nil ->
+        {:reply, :no_expectation, state}
```

Comment:
I wonder if we should break this in two checks: no_expectation and unknown_process. Unknown process means no expectation was ever defined for this PID, no_expectation means the pid is known, but that particular key was not mocked. Thoughts?

Reply:
We can also postpone this until we add multiple processes because that's when this distinction will matter the most!